### DISCUSSION TITLE
how to setup devops as private datasource for renovate?

### DISCUSSION THREAD

**Discussion by marvinfischenich** [renovatebot/renovate#16418] [Request Help]
Hi guys,
is there a docu how to setup devops to be used as private datasource for renovate bot?
Currently i get
`DEBUG: Found no results from datasource that look like a version (repository=apps/base-image-update/base-image-update, dependency=mypython-app)
       "result": {"releases": [], "isPrivate": true}
`
Can renovate bot consume a init.py file from the dependency?
Does it consider tags?

the renovate.json of the repository looks like this:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "github>renovatebot/.github"
  ],
  "includeForks": true,
  "platformAutomerge": true,
  "platform": "gitlab",
  "hostRules": [
    {
      "matchHost": "https://dev.azure.com/myorg/mypython-app/",
      "token": "abc"
    },
    {
      "matchHost": "https://dev.azure.com/myorg/mypython-app/_git/",
      "token": "abc"
    }
  ],
  "packageRules": [
    {
      "matchDatasources": "pypi",
      "matchPackagePatterns": [
        "^mypython-app"
      ],
      "registryUrls": [
        "https://dev.azure.com/myorg/mypython-app",
        "https://dev.azure.com/myorg/mypython-app/_git/"
      ]
    }
  ]
}
```

Regards, Marvin

**Comment by rarkins** [comment]
You will see that log message if (a) the lookup succeeded, (b) there was at least one result, (c) none of the results satisfied versioning.isVersion(). e.g. none looked like a valid PyPI version string. It does not indicate that there's any problem with credentials or private.

So what do the returned versions look like?

**Reply by marvinfischenich** [reply]
what do you mean by "returned version"? the dependency lookup?

```
             "deps": [
               {
                 "depName": "mypython-app",
                 "currentValue": "== 1.5.21",
                 "datasource": "pypi",
                 "currentVersion": "1.5.21",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [],
                 "versioning": "pep440"
               }
             ]
```

**Reply by rarkins** [reply]
What versions are on that datasource as part of the lookup result?

**Reply by marvinfischenich** [reply]
that depends on what renovate detects as a version.
that was part of my initial question is there a documentation about what renovate detects as a version when consuming devops as datasource?
i couldn't find any.

does it require Tags in devops?
then we have several versions (also formats for testing) like v.1.5.20, v1.3.18, v1.3.23rc1

or is renovate looking into main branch and into a specific file like an __init__.py which consists of
```
"""Initializing the mypyhton-app library"""
__version__ = "1.5.29"

```

**Reply by rarkins** [reply]
Can you clarify in more detail what you mean by "devops as private datasource" and "devops as datasource"?

**Reply by marvinfischenich** [reply]
Sorry, i meant the same. 
I have a private devops repo inside our organization where the dependency is stored which renovate should check for available updates.

**Reply by rarkins** [reply]
Usually when we say repo we are talking about files and git. When we use the term "registry" we're talking about packages and dependencies. How exactly are the "available updates" stored within the *repository*? Do you have any public example?

**Reply by marvinfischenich** [reply]
I have a pip.ini file on my local machine:
```
[global]
index-url=https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/
```

When i run `pip install mydependency` in that folder the python package gets installed successfully.
So i don't get why renovate isn't finding any version on DevOps.

I have tried to set the URL above as index-url in the requirements.txt, as package and as host-rule.
But nothing seems to work.

Where do i have to maintain that package url?

**Reply by viceice** [reply]
you need a package rule for pypi datasource wit a registry url set to that exact same pip index url

**Comment by marvinfischenich** [comment]
my renovate.json looks like this:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": ["github>renovatebot/.github"],
  "includeForks": true,
  "platformAutomerge": true,
  "platform": "gitlab",
  "hostRules": [
    {
      "hostType": "azure",
      "matchHost": "https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/",
      "token": "..."
    }
  ],
  "packageRules": [
    {
      "matchDatasources":"pypi",
      "matchPackagePatterns": ["^mydependency"],
      "registryUrls":[
        "https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/"
      ]
    }
  ]
}
```

the requirements.txt:
```
--index-url https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/
mydependency == 1.5.21
```

do you see any misconfiguration there?
i get

```
DEBUG: Found 2 package file(s) (repository=apps/base-image-update/base-image-update)
 INFO: Dependency extraction complete (repository=apps/base-image-update/base-image-update)
       "baseBranch": "master",
       "stats": {
         "managers": {
           "gitlabci-include": {"fileCount": 1, "depCount": 1},
           "pip_requirements": {"fileCount": 1, "depCount": 1}
         },
         "total": {"fileCount": 2, "depCount": 2}
       }
DEBUG: Datasource unauthorized (repository=apps/base-image-update/base-image-update)
       "datasource": "pypi",
       "packageName": "mydependency",
       "url": "https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/mydependency/"
DEBUG: Failed to look up dependency mydependency (repository=apps/base-image-update/base-image-update, packageFile=requirements.txt, dependency=mydependency)
```

the token has to be correct. i used the same on my local machine.

**Reply by viceice** [reply]
remove hostType from host rule. if any, it needs to match `pypi` datasource name instead

**Reply by viceice** [reply]
@HonkingGoose we probably should add some details about auth for datasource to docs. by default it needs to match the datasource id ( internal name). 🤔

**Reply by marvinfischenich** [reply]
i removed the hosttype but am still getting

```
DEBUG: Found 2 package file(s) (repository=apps/base-image-update/base-image-update)
 INFO: Dependency extraction complete (repository=apps/base-image-update/base-image-update)
       "baseBranch": "master",
       "stats": {
         "managers": {
           "gitlabci-include": {"fileCount": 1, "depCount": 1},
           "pip_requirements": {"fileCount": 1, "depCount": 1}
         },
         "total": {"fileCount": 2, "depCount": 2}
       }
DEBUG: Datasource unauthorized (repository=apps/base-image-update/base-image-update)
       "datasource": "pypi",
       "packageName": "mydependency",
       "url": "https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/mydependency/"
DEBUG: Failed to look up dependency mydependency (repository=apps/base-image-update/base-image-update, packageFile=requirements.txt, dependency=mydependency)
```


```
         "pip_requirements": [
           {
             "packageFile": "requirements.txt",
             "deps": [
               {
                 "depName": "mydependency",
                 "currentValue": "== 1.5.21",
                 "datasource": "pypi",
                 "currentVersion": "1.5.21",
                 "depIndex": 0,
                 "updates": [],
                 "warnings": [
                   {
                     "topic": "mydependency",
                     "message": "Failed to look up dependency mydependency"
                   }
                 ],
                 "versioning": "pep440"
               }
             ],
             "registryUrls": [
               "https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/"
             ]
           }
         ]
```


```
DEBUG: http statistics (repository=apps/base-image-update/base-image-update)
       "urls": {
         "https://api.github.com/repos/renovatebot/.github/contents/default.json (GET,200)": 1,
         "https://api.github.com/repos/whitesource/merge-confidence/contents/beta.json (GET,200)": 1,
         "https://git.diap.mblb.net/api/v4/projects/apps%2Fbase-image-update%2Fbase-image-update/issues (GET,200)": 1,
         "https://git.diap.mblb.net/api/v4/projects/apps%2Fbase-image-update%2Fbase-image-update/merge_requests (GET,200)": 1,
         "https://git.diap.mblb.net/api/v4/projects/apps%2Fbase-image-update%2Fbase-image-update/repository/files/renovate.json (GET,200)": 1,
         "https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/mydependency/ (GET,401)": 1
       },
```

i think the GET,401 is interesting. don't know why that is happening.

**Reply by viceice** [reply]
yes, authentication doesn't seem to work.

Try to `curl -H "authorization: Bearer <token>" https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/mydependency/` 

How do you authenticate to azure from your local machine?

**Please don't use `github>renovatebot/.github` preset, otherwise renovate maintainers get flooded by your renovate PR's!**

**Reply by marvinfischenich** [reply]
> Try to `curl -H "authorization: Bearer <token>" https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/mydependency/`

that's not working i receive
`{"$id":"1","innerException":null,"message":"TF400813: The user '' is not authorized to access this resource.","typeName":"Microsoft.TeamFoundation.Framework.Server.UnauthorizedRequestException, Microsoft.TeamFoundation.Framework.Server","typeKey":"UnauthorizedRequestException","errorCode":0,"eventId":3000}`

however
`curl https://pkgs.dev.azure.com/myorg/mydependency/_packaging/mydependency/pypi/simple/mydependency/ -u my.user@myorg.com:token`
is working. here i receive the html.

> How do you authenticate to azure from your local machine?

pip is asking for user and password. i use the same credentials as in curl.

**Reply by viceice** [reply]
configure username and password in your host rule instead of token

**Reply by marvinfischenich** [reply]
that did the trick, thx!

```
                    "bucket": "non-major",
                     "newVersion": "1.5.29",
                     "newValue": "==1.5.29",
                     "newMajor": 1,
                     "newMinor": 5,
                     "updateType": "patch",
                     "isRange": true,
                     "branchName": "renovate/mydependency-1.x"
```
